src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryModule.java

```diff
@@ -112,6 +112,10 @@ public StarlarkCallable repositoryRule(
     BazelModuleContext moduleContext = BazelModuleContext.ofInnermostBzlOrThrow(thread);
     builder.setRuleDefinitionEnvironmentLabelAndDigest(
         moduleContext.label(), moduleContext.bzlTransitiveDigest());
+    Label.RepoMappingRecorder repoMappingRecorder = thread.getThreadLocal(Label.RepoMappingRecorder.class);
```

Comment:
I have; the linked bug focuses on a different topic. I think it shouldn't matter for repo rules used in bzlmod anyway, which have to be exported symbols and thus cannot close over anything lower in the call stack.

Reply:
actually, this _is_ a problem: I talked to @brandjon about this, and he explained to me what's wrong here. Imagine the following:

```starlark
# foo.bzl
def make_repo_rule(impl):
    return repository_rule(impl)

# bar.bzl
load(":foo.bzl", "make_repo_rule")
def _my_rule_impl(rctx):
   # ...
my_rule = make_repo_rule(_my_rule_impl)
```

With the current code, the "transitive bzl digest" of `my_rule` is actually the transitive bzl digest of `foo.bzl`, since that's where the `repository_rule()` call happens. What we _should_ obviously be using is `bar.bzl`, which is actually at the bottom of the call stack. In order to do that, line 112 should simply be grabbing the information from the corresponding BzlInitThreadContext (née BazelStarlarkContext). I'll send a separate fix for this.

In any case, this PR itself is not affected; we're already using the "repo mapping recorder" from the "bottom of the call stack", since there's only one recorder for the entire thread, not one per call stack frame.

Reply:
actually, you know what... I'll just fix it in this PR.

Reply:
gaah.. I reverted the fix. The reason is that we current allow usage of unexported repo rules in WORKSPACE, like:

```starlark
# foo.bzl
def foo()
    abc = repository_rule(...)
    abc(name="def")

# WORKSPACE
load(":foo.bzl", "foo")
foo()
```

In this case, there is no BzlInitThreadContext for us to work with, since the "bottom of the stack" is literally the WORKSPACE file itself.

So properly fixing this TODO would break WORKSPACE usage. I'll send a separate change to gate the fix behind `--noenable_workspace` instead.

Reply:
hmm, that's true. We can check for existence of BzlInitThreadContext. If it exists, do the proper thing; otherwise, fall back to the current wrong behavior. Will address in another change anyway.